clusterprovider/kcp/kcp.go

```diff
@@ -0,0 +1,255 @@
+	return k.clientset.Delete(k.ctx, name, metav1.DeleteOptions{})
+}
+
+// ListClustersNames is N/A for KCP
+// TODO remove it from ProviderClient interface
+func (k KcpClusterProvider) ListClustersNames() ([]string, error) {
+	return []string{}, nil
+}
+
+func (k *KcpClusterProvider) Get(lcName string) (clusterprovider.LogicalClusterInfo, error) {
+	cfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: k.configPath},
+		&clientcmd.ConfigOverrides{CurrentContext: "root"}).RawConfig()
+	if err != nil {
+		k.logger.Error(err, "Failed to get config for workspace", "workspace", lcName)
+	}
+
+	cfgBytes, err := clientcmd.Write(buildRawConfig(cfg, lcName))
+	if err != nil {
+		k.logger.Error(err, "Failed to write workspace config", "workspace", lcName)
+	}
+
+	lcInfo := clusterprovider.LogicalClusterInfo{
+		Name:   lcName,
+		Config: string(cfgBytes[:]),
+	}
+	return lcInfo, err
+}
+
+func (k *KcpClusterProvider) ListClusters() ([]clusterprovider.LogicalClusterInfo, error) {
+	list, err := k.clientset.List(k.ctx, metav1.ListOptions{})
+	if err != nil {
+		return []clusterprovider.LogicalClusterInfo{}, err
+	}
+	lcInfoList := make([]clusterprovider.LogicalClusterInfo, 0, len(list.Items))
+
+	for _, ws := range list.Items {
+		lcInfo, err := k.Get(ws.Name)
+		if err != nil {
+			k.logger.Error(err, "Failed to fetch config for workspace", "name", ws.Name)
+		}
+
+		lcInfoList = append(lcInfoList, lcInfo)
+	}
+	return lcInfoList, nil
+}
+
+func (k *KcpClusterProvider) Watch() (clusterprovider.Watcher, error) {
+	w := &KcpWatcher{
+		ch:       make(chan clusterprovider.WatchEvent),
+		provider: k}
+	k.watch = w
+	return w, nil
+}
+
+type KcpWatcher struct {
+	init     sync.Once
+	wg       sync.WaitGroup
+	ch       chan clusterprovider.WatchEvent
+	cancel   context.CancelFunc
+	provider *KcpClusterProvider
+}
+
+func (k *KcpWatcher) Stop() {
+	if k.cancel != nil {
+		k.cancel()
+	}
+	k.wg.Wait()
+	close(k.ch)
+}
+
+func (k *KcpWatcher) ResultChan() <-chan clusterprovider.WatchEvent {
+	k.init.Do(func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		k.cancel = cancel
+
+		k.wg.Add(1)
+		go func() {
+			defer k.wg.Done()
+			watcher, err := k.provider.clientset.Watch(k.provider.ctx, metav1.ListOptions{})
+			if err == nil {
+				for {
+					select {
+					case <-ctx.Done():
+						watcher.Stop()
+						k.Stop()
+					case event, ok := <-watcher.ResultChan():
+						if !ok {
+							k.Stop()
+							return
+						}
+						ws := event.Object.(*tenancyv1alpha1.Workspace)
+						name := ws.Name
+						if event.Type == "MODIFIED" {
+							k.provider.logger.V(2).Info("KCP workspace modify event", "ws", ws.Name)
+							_, ok := k.provider.workspaces[name]
+							// check ws is ready
+							if ws.Status.Phase != corev1alpha1.LogicalClusterPhaseReady {
+								k.provider.logger.V(2).Info("KCP workspace is not ready")
+								if ok {
```

Comment:
Maybe move the if OK to before if ws.Status ? seems more readable..  and the the other section is else

Reply:
Changed